gen/src/generate/idna/mapping.rs

```diff
@@ -0,0 +1,143 @@
+use std::char;
+use std::fs::File;
+use std::io::{self, Read, Write};
+use std::path::Path;
+use std::str::FromStr;
+
+use super::UnicodeVersion;
+
+use regex::Regex;
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+struct IdnaDataLine {
+    low: char,
+    high: char,
+
+    /// valid, ignored, mapped, deviation, disallowed,
+    /// disallowed_STD3_valid, or disallowed_STD3_mapped
+    status: String,
+
+    /// Only present if the status is ignored, mapped, deviation, or disallowed_STD3_mapped.
+    mapping: Option<Box<[char]>>,
+
+    /// There are two values: NV8 and XV8.
+    /// NV8 is only present if the status is valid but the character is excluded
+    /// by IDNA2008 from all domain names for all versions of Unicode.
+    /// XV8 is present when the character is excluded
+    /// by IDNA2008 for the current version of Unicode.
+    /// These are not normative values.
+    idna_2008_status: Option<String>,
+}
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+struct IdnaMapping(Box<[IdnaDataLine]>);
+
+impl IdnaMapping {
+    fn emit<P: AsRef<Path>>(&self, dir: P) -> io::Result<()> {
+        let mut file = File::create(dir.as_ref().join("idna_mapping.rsv"))?;
+        let IdnaMapping(ref lines) = *self;
+        writeln!(file, "&[")?;
+        for line in lines.iter() {
+            write!(
+                file,
+                "    ('{}', '{}', {}",
+                line.low.escape_unicode(),
+                line.high.escape_unicode(),
+                match line.status.as_str() {
+                    "valid" => "Valid",
+                    "ignored" => "Ignored",
+                    "mapped" => "Mapped",
+                    "deviation" => "Deviation",
+                    "disallowed" => "Disallowed",
+                    "disallowed_STD3_valid" => "DisallowedStd3Valid",
+                    "disallowed_STD3_mapped" => "DisallowedStd3Mapped",
+                    _ => unreachable!(),
+                },
+            )?;
+            if let Some(ref chars) = line.mapping {
+            write!(file, "(&[")?;
+                for char in chars.iter() {
+                    write!(file, "'{}',", char.escape_unicode())?;
+                }
+            write!(file, "])")?;
+            }
+            writeln!(file, "),")?;
+        }
+        writeln!(file, "]")?;
+        Ok(())
+    }
+}
+
+impl FromStr for IdnaMapping {
+    type Err = ();
+
+    fn from_str(str: &str) -> Result<Self, Self::Err> {
+        lazy_static! {
```

Comment:
Scoping it in the function keeps the declaration close to the use site. [It's also how the regex crate suggests avoiding recompilation of the regex.](https://doc.rust-lang.org/regex/regex/index.html#example-avoid-compiling-the-same-regex-in-a-loop)

I could bring it out, but that would increase the visibility and I don't see any need to, especially since this follows the example on avoiding recompilation of regex.